### DISCUSSION TITLE
Making it Easier to Input Special Characters

### DISCUSSION THREAD

**Discussion by Glitchy-Tozier** [8VIM/8VIM#110] [Ideas]
Let's have a discussion about the symbols.

- the uppercase-@-shortcut should be indicated somewhere. Also, I'm not sure whether it should exist. Maybe placing a # there might be more useful. Personally, I rarely find myself thinking "man, I wish i had a "@gmail.com"-button.
- I would much rather the lowercase special symbol be `"` instead of `'`.
- The 2-turn thing with `!` is fun (`YO!! _\m/`) but I'm not sure what its use is.
- Why is ! the same in uppercase?

**Comment by m-rey** [comment]
> the uppercase-@-shortcut should be indicated somewhere. Also, I'm not sure whether it should exist. Maybe placing a # there might be more useful. Personally, I rarely find myself thinking "man, I wish i had a "@gmail.com"-button.

Totally agree. I didn't even notice it. Having a fixed `@gmail.com` is not that useful, compared to having another character like `#` or `&`.

> I would much rather the lowercase special symbol be " instead of '.

I disagree. Having `'` instead of `"` allows for easier typing of words that contain it, like `it's` or `can't`. It's often used in English. Surely more often than `"`.

> The 2-turn thing with ! is fun (YO!! _\m/) but I'm not sure what its use is.

I guess it's an easter egg. I never stumbled upon it, without checking the code. I'm indifferent whether it should be kept or not. I guess that it doesn no harm to keep it for now.

>  Why is ! the same in uppercase?

This could be replaced by another character, indeed. Again, `#` or `&` come to mind.

**Reply by geofrey** [reply]
\m|,  YO!!  ,|m/

**Comment by Glitchy-Tozier** [comment]
> I disagree. Having `'` instead of `"` allows for easier typing of words that contain it, like `it's` or `can't`. It's often used in English. Surely more often than `"`.

Very good point, didn't think of that. (I'm mainly writing in German.)

> This could be replaced by another character, indeed. Again, `#` or `&` come to mind.

Exactly, that were the symbols I had in mind.
However, I think we should only change this when special character inputting is improved (see https://github.com/flide/8VIM/issues/99#issuecomment-733818251). Currently, I like the thought of WRITING SOMETHING WITH CAPSLOCK AND BEING ABLE TO EASILY ADD EXCLAMATION MARKS WITHOUT NEEDING TO DISABLE CAPS-LOCK!!!!!!!

**Comment by sslater11** [comment]
I feel like this is a big topic to cover, because there will be so many personal opinions.
I think we should have a second layout similar to the 8pen layout, but with symbols. That being said, there are a lot of different symbols, and also numbers too!

I guess numbers should be another separate one, but then we get the question of how we can quickly switch between these. I also think certain symbols should be duplicated on the numbers layout, because I _always_ wish for an easy way to type 12:15. No matter what keyboard I've used, typing in a time or date has always felt clunky.

The problem I see is that all of this leads to having to learn more, giving vim8 a higher learning curve, or just another keyboard with a clunky feeling of "oh where'd they put that key".

I think grouping symbols together is a must too.
In this example the brackets and maths bit make perfect sense to everyone, but the bitwise operators only makes perfect sense to a programmer, and so only programmers would easily remember where those are.

Obviously this is just a rubbish sketch up and not a real layout.
It's also missing comma and full stop, which I think both deserve their own way of typing, but that's probably just me.
![vim8-symbols-idea](https://user-images.githubusercontent.com/43177940/100278996-a31d4300-2f5d-11eb-8b6d-bbe53648ecce.jpg)

**Reply by Treeoctopus** [reply]
yes! I'd love this -- switching over to a hunt and peck keyboard really does my head in

**Comment by Glitchy-Tozier** [comment]
> Obviously this is just a rubbish sketch up and not a real layout.

I actually really like them. You make them look so much less caotic than the ones I create (https://github.com/flide/8VIM/issues/90#issuecomment-732778061) I especially like the visuals of having grouped brackets.

> I think we should have a second layout similar to the 8pen layout, but with symbols. That being said, there are a lot of different symbols, and also numbers too!

However, I'm not sure another X-layout is necessary. We already have these two keyboards https://github.com/flide/8VIM/issues/77#issuecomment-733604781 (you can open these by clicking on the left sector) and I think they should do a pretty good job. (I haven't used this keyboard a lot yet BECAUSE THERE STILL IS NO GERMAN LAYOUT!!)
I think the main issue is that we need a better way to input the **really common** special characters. For example, I rarely type `@`, `[`, `]`, `{`, `}`, `=`, `%`, or `|`.
Additionally, for characters in the outer layers of the special-characters-X-board, it might be faster to open the special-characters-keyboard.
For the most important special characters we need a better input method though.

This might not solve all our problems, but I think something like **this** may also be worth considering(ignore the letters: it's a draft for a German layout)
![8vimSpecial](https://user-images.githubusercontent.com/59611881/100332653-f03afc80-2fd1-11eb-8a04-64ed16d84bd5.jpeg)


I placed CTRL on the left so that the keyboard doesn't have that weird margin at the bottom anymore. @flide, please to something about this some day. It's not visually appealing.
Also, WOOOW, I just had an Idea. Those buttons could be activated by the user by swiping from the circle to the button of your choice. This would click the button to insert the special character and (possibly) integrate nicely into your writing-flow.

What I like about this idea:
1. Finally we can have the same margin on the top & bottom of the X-Keyboard as `CTRL` now is on the left
2. All the buttons that you would use "while typing" now are on the left.
3. Obviously we have easy access to some very common special characters.

Problems with this idea:
1. The main problem with this seems to me that there are too few symbols that can be fit in there. (For many languages, the comma `,` is important. English needs a `'` and I people migh want access to at least one other special character of their choosing.
2. Screen size might start getting limited.

**Comment by flide** [comment]
We can surely replace symbols on the xpad, but it would be better to have a concrete resoning behind any change. 

Side note : if it's going to be a change on whim, I assure you, my laziness triumps all.

**Comment by flide** [comment]
No xpad style board for numbers or symbols, because none of them are frequent enough to learn them. Also, I can bet that the 10 digits [0-9] are pretty evenly distributed in the texts of ANY language (42 might be an exception, but then it's too special to be bound by any rules). 

8Pen used to have xpad style boards for symbols and numbers, frustrated me to no end. It's like learning the keyboard everytime i want to type in numbers. 

Strict NO for xpad style number/symbol board.

**Comment by Glitchy-Tozier** [comment]
@flide I absolutely agree that we have to properly think through any changes.

**Comment by sslater11** [comment]
@flide, as I said that there may be a higher learning curve, and well you sound like you have more experience/frustration, so I will agree with you. No xpad for numbers and symbols.

I would like a quick way to input certain most commonly used symbols, but I think everyone uses different ones, and there's a surprising amount of common symbols.
Here's my personal selection of most useful ones ( ) ! ? , . : - / " ' £ * all of which I use in normal texts. Would adding all these not clutter up the keyboard area?

Maybe in the future gestures could fix this? e.g. I can draw an open bracket and wow it's there.

**Comment by m-rey** [comment]
> We can surely replace symbols on the xpad, but it would be better to have a concrete reasoning behind any change.

I propose replacing the `@gmail.com` snippet of the "uppercase" `@` for a few reasons:
* there are other more common characters in normal conversation than `@gmail.com` that are currently missing in the main view
* it's not documented anywhere (yet?)
* hardly anyone knew it was there, meaning that is used very rarely, if at all 
* it's confusing, as it's not clear that it expands to `@gmail.com`
* it's confusing, as the same symbol represents different things in uppercase and lowercase

I'd replace it with another character or alternatively would offer a way of changing the `@gmail.com` text to something different. (in settings maybe?)

I'd also propose replacing the uppercase or lowercase `!` with a different character.
* one character occupying two slots prevents another character from being on the xpad
* it's confusing. After observing that while both uppercase and lowercase `@` use the same symbol, they result in different results, users may now think that there may be a difference between the uppercase and lowercase `!`. But there is none.

**Comment by Glitchy-Tozier** [comment]
I agree with you on `@`.

But `!` is debatable in my opinion.
- If we improve inputting special characters, it's fine to replace `uppercase !` with something else.
- But as it currently stands, I wouldn't want to loose the ability to CAPITALIZE AND THEN WRITE EASY EXCLAMATION MARKS!!

**Comment by flide** [comment]
@Glitchy-Tozier maybe we can have '!' as uppercase and change the character when in lowercase, to allow ppl have there shouting matches without interuption.

**Comment by Glitchy-Tozier** [comment]
@flide I thought about the same thing, but there's one problem: People unfortunately also use `!` in "regular" circumstances. And removing lowercase `!` makes it really inaccessible.

**Comment by flide** [comment]
@Glitchy-Tozier not really, it will still be accessible with a double circle like capital letters. 

Also, don't overthink it. there exists only 32 positions and some characters are bound to be left behind. Let go of them Glitchy!

**Comment by Glitchy-Tozier** [comment]
1. See https://github.com/flide/8VIM/issues/115#issuecomment-734876176 for another idea.

2. >not really, it will still be accessible with a double circle like capital letters.

Yes, but that's not a productive thing to do. I understand that you would do that to something like `#` or `@`, but `!`?? It's one of the three ways to finish a sentence. I guarantee that (in non-business-texts) it's way more common than something like `z` in English.

It's too important to lock away behind **2 turns**. If we _have_ to choose one, keep it at lowercase. I'm just saying that i wouldn't mind having it in BOTH, lower- and uppercase.

**Comment by m-rey** [comment]
> It's too important to lock away behind 2 turns. If we have to choose one, keep it at lowercase.

Agreed, please keep it in lowercase. If you are raging enough to write in full caps, you are able to make 2 turns to finish it off with an `!` ;)

**Comment by caseyng** [comment]
bringing my discussion here. Is there any way to bring the numbers into the main x pad? Cuz it's clunky to have to switch between numbers and alphabets every time I want to type numbers in the middle of a sentence.

**Comment by flide** [comment]
guys join in on matrix channel, it will be earier to converse.

https://matrix.to/#/+8vim:matrix.org

**Comment by BruceMustache** [comment]
> @flide, as I said that there may be a higher learning curve, and well you sound like you have more experience/frustration, so I will agree with you. No xpad for numbers and symbols

I thing xpads should be implemented wherever be possible since not doing this totaly breaks this keyboard writing-flow

**Comment by flide** [comment]
@BruceMustache I will try to explain. 

The xpad writing style is more or less like muscle memory, like how our hands 'know' how to write a letter on paper. You don't have to think how to write 'b' on paper, you think and it happens. 

Same thing happens with xpad. Once you know it, well... you know it. 
Now the problem, When there are more then 1 systems of xpad to remember, the hand no longer knows it and memory is needed to figure out which loops symbolize what character. This slows you down on the xpad you frequently use and makes you peek and refresh your memory when using the infrequent board. 

This is the major reason I was lobbying for a single script board rather than a board for each language, but I conceded since there exist only few multilinguals and the problem would not be that severe. 

Above was the first major problem. The second one is about frequency and loop assignment for numbers and symbols. Numbers are absolutely evenly distributed in terms of frequency in text, similarly for symbols (not an even distribution but, i believe, the difference will not be significant). So the practicality would state that someone needs to spray the board with chosen 32 symbols and numbers and assign them loops, and everybody got to remember them. 

At this point, you have to trust me it's not worth it. If you still disagree, you are free to fork the code and develop that number/symbols board for testing.

**Comment by Glitchy-Tozier** [comment]
So how about this?
 ![8vimSpecial](https://user-images.githubusercontent.com/59611881/100332653-f03afc80-2fd1-11eb-8a04-64ed16d84bd5.jpeg)

**Reply by flide** [reply]
I was thinking maybe we should have a repository for the custom keyboards being developed, so that if someone wants to use them they can just download it and using the iteration 8 load and use it.

**Reply by Glitchy-Tozier** [reply]
Sounds like a decent idea!
